File: zip-project.ts
import AdmZip from 'adm-zip';
import { execSync } from 'child_process';
import * as fs from 'fs';
import * as path from 'path';

async function main() {
  const rootDir = process.cwd();
  const publicDir = path.join(rootDir, 'public');
  const distDir = path.join(rootDir, 'dist');

  // public ディレクトリがなければ作成
  if (!fs.existsSync(publicDir)) {
    fs.mkdirSync(publicDir, { recursive: true });
    console.log('📁 public ディレクトリを作成しました。');
  }

  const releaseZipName = 'tobita-girls-website-release.zip';
  const sourceZipName = 'tobita-girls-source-code.zip';

  const rootReleaseZip = path.join(rootDir, releaseZipName);
  const rootSourceZip = path.join(rootDir, sourceZipName);
  
  const publicReleaseZip = path.join(publicDir, releaseZipName);
  const publicSourceZip = path.join(publicDir, sourceZipName);

  // 1. 古いZIPを削除
  [rootReleaseZip, rootSourceZip, publicReleaseZip, publicSourceZip].forEach(p => {
    if (fs.existsSync(p)) fs.unlinkSync(p);
  });

  // ----------------------------------------------------
  // ① まず一時的に Production ビルドを一度実行して dist を作ります
  // ----------------------------------------------------
  console.log('📦 1. 一時Productionビルドを実行中...');
  try {
    execSync('npm run build', { stdio: 'inherit' });
  } catch (error) {
    console.error('❌ ビルド中にエラーが発生しました:', error);
  }

  // ----------------------------------------------------
  // ② 静的ウェブサイトZIP (dist フォルダ内容) の作成
  // ----------------------------------------------------
  console.log('📦 2. 本番用ウェブサイトZIP (tobita-girls-website-release.zip) を作成中...');
  if (fs.existsSync(distDir)) {
    const zipRelease = new AdmZip();
    const distItems = fs.readdirSync(distDir);
    for (const item of distItems) {
      if (item.endsWith('.zip')) continue;
      const fullPath = path.join(distDir, item);
      const stat = fs.statSync(fullPath);
      if (stat.isDirectory()) {
        zipRelease.addLocalFolder(fullPath, item);
      } else {
        zipRelease.addLocalFile(fullPath);
      }
    }
    
    zipRelease.writeZip(rootReleaseZip);
    fs.copyFileSync(rootReleaseZip, publicReleaseZip);
    fs.copyFileSync(rootReleaseZip, path.join(distDir, releaseZipName));
    console.log(`✅ 作成完了: ${rootReleaseZip} (${(fs.statSync(rootReleaseZip).size / (1024 * 1024)).toFixed(2)} MB)`);
  } else {
    console.error('❌ distディレクトリが見つかりません。');
  }

  // ----------------------------------------------------
  // ③ ソースコード全体のZIP作成
  // ----------------------------------------------------
  console.log('📦 3. ソースコード一式ZIP (tobita-girls-source-code.zip) を作成中...');
  const zipSource = new AdmZip();
  
  const items = fs.readdirSync(rootDir);
  const excludeList = [
    'node_modules',
    'dist',
    '.git',
    '.github',
    releaseZipName,
    sourceZipName,
    '.DS_Store'
  ];

  for (const item of items) {
    if (excludeList.includes(item) || item.endsWith('.zip')) continue;
    
    const fullPath = path.join(rootDir, item);
    const stat = fs.statSync(fullPath);
    
    if (stat.isDirectory()) {
      zipSource.addLocalFolder(fullPath, item);
    } else {
      zipSource.addLocalFile(fullPath);
    }
  }

  zipSource.writeZip(rootSourceZip);
  fs.copyFileSync(rootSourceZip, publicSourceZip);
  fs.copyFileSync(rootSourceZip, path.join(distDir, sourceZipName));
  console.log(`✅ 作成完了: ${rootSourceZip} (${(fs.statSync(rootSourceZip).size / (1024 * 1024)).toFixed(2)} MB)`);

  console.log('\n 🎉 すべてのZIPファイルの生成に成功しました！');
}

main().catch(err => {
  console.error('Unexpected error:', err);
  process.exit(1);
});

File: zip-project.py
import os
import shutil
import subprocess
import sys
import zipfile

def create_zip_from_dir(dir_path, zip_path):
    print(f"  -> Creating {zip_path} from folder {dir_path}...")
    with zipfile.ZipFile(zip_path, 'w', zipfile.ZIP_DEFLATED) as zip_file:
        for root, dirs, files in os.walk(dir_path):
            # Sort for determinism
            dirs.sort()
            files.sort()
            
            # Create directory entries
            for d in dirs:
                full_dir_path = os.path.join(root, d)
                rel_dir_path = os.path.relpath(full_dir_path, dir_path)
                rel_dir_path = rel_dir_path.replace(os.path.sep, '/') + '/'
                
                info = zipfile.ZipInfo(rel_dir_path)
                # UNIX directory permissions 755
                info.external_attr = (0o40755 << 16) | 0x10
                zip_file.writestr(info, '')
                
            # Create file entries
            for file in files:
                full_path = os.path.join(root, file)
                rel_path = os.path.relpath(full_path, dir_path)
                rel_path = rel_path.replace(os.path.sep, '/')
                
                info = zipfile.ZipInfo(rel_path)
                # UNIX file permissions 644
                info.external_attr = (0o100644 << 16)
                info.compress_type = zipfile.ZIP_DEFLATED
                
                with open(full_path, 'rb') as f:
                    zip_file.writestr(info, f.read())

def create_source_zip(root_dir, zip_path, exclude_list):
    print(f"  -> Creating {zip_path} for source code in {root_dir}...")
    with zipfile.ZipFile(zip_path, 'w', zipfile.ZIP_DEFLATED) as zip_file:
        for root, dirs, files in os.walk(root_dir):
            # Exclude folders
            dirs[:] = [d for d in dirs if d not in exclude_list]
            dirs.sort()
            files.sort()
            
            # Create directory entries
            for d in dirs:
                full_dir_path = os.path.join(root, d)
                rel_dir_path = os.path.relpath(full_dir_path, root_dir)
                rel_dir_path = rel_dir_path.replace(os.path.sep, '/') + '/'
                
                info = zipfile.ZipInfo(rel_dir_path)
                # UNIX directory permissions 755
                info.external_attr = (0o40755 << 16) | 0x10
                zip_file.writestr(info, '')
                
            # Create file entries
            for file in files:
                if file in exclude_list:
                    continue
                
                full_path = os.path.join(root, file)
                rel_path = os.path.relpath(full_path, root_dir)
                rel_path = rel_path.replace(os.path.sep, '/')
                
                # Check segments for safety
                segments = rel_path.split('/')
                if any(seg in exclude_list for seg in segments):
                    continue
                    
                info = zipfile.ZipInfo(rel_path)
                # UNIX file permissions 644
                info.external_attr = (0o100644 << 16)
                info.compress_type = zipfile.ZIP_DEFLATED
                
                with open(full_path, 'rb') as f:
                    zip_file.writestr(info, f.read())

def main():
    root_dir = os.getcwd()
    public_dir = os.path.join(root_dir, 'public')
    dist_dir = os.path.join(root_dir, 'dist')
    
    if not os.path.exists(public_dir):
        os.makedirs(public_dir, exist_ok=True)
        print("📁 Created public directory.")
        
    release_zip_name = 'tobita-girls-website-release.zip'
    source_zip_name = 'tobita-girls-source-code.zip'
    
    root_release_zip = os.path.join(root_dir, release_zip_name)
    root_source_zip = os.path.join(root_dir, source_zip_name)
    public_release_zip = os.path.join(public_dir, release_zip_name)
    public_source_zip = os.path.join(public_dir, source_zip_name)
    
    # 1. Delete old ZIP files
    for p in [root_release_zip, root_source_zip, public_release_zip, public_source_zip]:
        if os.path.exists(p):
            os.remove(p)
            print(f"🗑️ Removed old ZIP: {p}")
            
    # 2. Run initial production build
    print("📦 1. Running initial production build to generate dist folder...")
    try:
        subprocess.run(["npm", "run", "build"], check=True)
        print("✅ Initial build successful.")
    except subprocess.CalledProcessError as e:
        print(f"❌ Build failed: {e}")
        sys.exit(1)
        
    # 3. Create release ZIP from dist folder
    print("📦 2. Creating release ZIP...")
    if os.path.exists(dist_dir):
        create_zip_from_dir(dist_dir, root_release_zip)
        shutil.copy2(root_release_zip, public_release_zip)
        print(f"✅ Release ZIP created: {root_release_zip} & {public_release_zip}")
    else:
        print("❌ Error: dist folder does not exist.")
        sys.exit(1)
        
    # 4. Create source ZIP
    print("📦 3. Creating source ZIP...")
    exclude_list = [
        'node_modules',
        'dist',
        '.git',
        '.github',
        release_zip_name,
        source_zip_name,
        '.DS_Store'
    ]
    create_source_zip(root_dir, root_source_zip, exclude_list)
    shutil.copy2(root_source_zip, public_source_zip)
    print(f"✅ Source ZIP created: {root_source_zip} & {public_source_zip}")
    
    # 5. Run final build to bundle everything
    print("📦 4. Running final production build to embed ZIPs...")
    try:
        subprocess.run(["npm", "run", "build"], check=True)
        print("✅ Final build completed. ZIP files are fully bundled and downloadable.")
    except subprocess.CalledProcessError as e:
        print(f"❌ Final build failed: {e}")
        sys.exit(1)
        
    print("\n🎉 All ZIP files generated successfully with perfect permissions!")

if __name__ == '__main__':
    main()
